src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/WeakEvent`1.cs

```diff
@@ -0,0 +1,131 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Immutable;
+using System.Runtime.CompilerServices;
+
+namespace Roslyn.Utilities;
+
+internal sealed class WeakEvent<TEventArgs>
+{
+    /// <summary>
+    /// Each registered event handler has the lifetime of an associated owning object. This table ensures the weak
+    /// references to the event handlers are not cleaned up while the owning object is still alive.
+    /// </summary>
+    /// <remarks>
+    /// Access to this table is guarded by itself. For example:
+    /// <code>
+    /// <see langword="lock"/> (<see cref="_keepAliveTable"/>)
+    /// {
+    ///   // Read or write to <see cref="_keepAliveTable"/>.
+    /// }
+    /// </code>
+    /// </remarks>
+    private readonly ConditionalWeakTable<object, EventHandler<TEventArgs>> _keepAliveTable = new();
```

Comment:
>new(); [](http://example.com/codeflow?start=94&length=6)

How much overhead is there for each CWT instance? Would there be benefit in making this static?

Reply:
➡️ It could result in some unexpected semantics if the table overlapped (e.g. removing the event handler from one event might remove the keep-alive status for a different event).

Note that in the current code, only one of these will exist (GlobalOptionService is a singleton).